File: lib/local-scripts/regenerate-thumbnails.php
<?php

// Replaces dashes in keys with underscores
foreach($args as $index => $arg) {
	$split = strpos($arg, "=");
	if ( $split ) {
		$key = str_replace('-', '_', substr( $arg , 0, $split ) );
		$value = substr( $arg , $split, strlen( $arg ) );

		// Removes unnecessary bash quotes
		$value = trim( $value,'"' ); 				// Remove last quote 
		$value = str_replace( '="', '=', $value );  // Remove quote right after equals

		$args[$index] = $key.$value;
	} else {
		$args[$index] = str_replace('-', '_', $arg);
	}

}

// Converts --arguments into $arguments
parse_str( implode( '&', $args ) );

// Loads CLI configs
$json = "{$_SERVER['HOME']}/.captaincore-cli/config.json";

if ( ! file_exists( $json ) ) {
	echo "Error: Configuration file not found.";
	return;
}

$config_data = json_decode ( file_get_contents( $json ) );
$system      = $config_data[0]->system;
$path        = $system->path;
$site        = ( new CaptainCore\Sites )->get( $site_id );
$environments = ( new CaptainCore\Site( $site_id ) )->environments();

foreach( $environments as $environment ) {
	$environment_name         = lcfirst( $environment->environment );
	$details                  = ( isset( $environment->details ) ? json_decode( $environment->details ) : (object) [] );
	if ( empty( $details->screenshot_base ) ) {
		echo "No captures found for {$site->site}-{$environment_name}\n";
		continue;
	}
	echo "Fetching latest capture screenshot for {$site->site}-{$environment_name}\n";
	$working_image    = "{$system->rclone_upload_uri}/{$captain_id}/{$site->site}_{$site->site_id}/{$environment_name}/captures/%23_{$details->screenshot_base}.jpg";
	echo shell_exec( "cd {$system->path_tmp}
wget -O {$site->site}_{$site->site_id}_{$details->screenshot_base}.jpg $working_image
convert {$site->site}_{$site->site_id}_{$details->screenshot_base}.jpg -thumbnail 800 -gravity North -crop 800x500+0+0 {$details->screenshot_base}_thumb-800.jpg
convert {$details->screenshot_base}_thumb-800.jpg -thumbnail 100 {$details->screenshot_base}_thumb-100.jpg
rclone move {$details->screenshot_base}_thumb-100.jpg {$system->rclone_upload}{$captain_id}/{$site->site}_{$site->site_id}/{$environment_name}/screenshots/ --fast-list --transfers=32 --no-update-modtime --progress
rclone move {$details->screenshot_base}_thumb-800.jpg {$system->rclone_upload}{$captain_id}/{$site->site}_{$site->site_id}/{$environment_name}/screenshots/ --fast-list --transfers=32 --no-update-modtime --progress
rm {$site->site}_{$site->site_id}_{$details->screenshot_base}.jpg");

}

File: cmd/regenerate-thumbnails.go
package cmd

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/CaptainCore/captaincore/models"
	"github.com/spf13/cobra"
)

var regenerateThumbnailsCmd = &cobra.Command{
	Use:   "regenerate-thumbnails <site>",
	Short: "Generate thumbnails from most recent capture",
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) < 1 {
			return errors.New("requires a <site> argument")
		}
		return nil
	},
	Run: func(cmd *cobra.Command, args []string) {
		resolveNativeOrWP(cmd, args, regenerateThumbnailsNative)
	},
}

// regenerateThumbnailsNative implements `captaincore regenerate-thumbnails <site>` natively in Go.
func regenerateThumbnailsNative(cmd *cobra.Command, args []string) {
	sa := parseSiteArgument(args[0])
	site, err := sa.LookupSite()
	if err != nil || site == nil {
		fmt.Printf("Error: Site '%s' not found.\n", sa.SiteName)
		return
	}

	environments, err := models.FindEnvironmentsBySiteID(site.SiteID)
	if err != nil {
		fmt.Printf("Error fetching environments: %v\n", err)
		return
	}

	_, system, captain, err := loadCaptainConfig()
	if err != nil || system == nil {
		fmt.Println("Error: Configuration file not found.")
		return
	}

	rcloneUpload := ""
	if captain != nil {
		rcloneUpload = captain.Remotes["rclone_upload"]
	}

	for _, env := range environments {
		envName := strings.ToLower(env.Environment)
		details := env.ParseDetails()

		if details.ScreenshotBase == "" {
			fmt.Printf("No captures found for %s-%s\n", site.Site, envName)
			continue
		}

		fmt.Printf("Fetching latest capture screenshot for %s-%s\n", site.Site, envName)

		siteDir := fmt.Sprintf("%s_%d", site.Site, site.SiteID)
		sourceURL := fmt.Sprintf("%s/%s/%s/%s/captures/%%23_%s.jpg", system.RcloneUploadURI, captainID, siteDir, envName, details.ScreenshotBase)
		rcloneDest := fmt.Sprintf("%s%s/%s/%s/screenshots/", rcloneUpload, captainID, siteDir, envName)
		downloadFile := fmt.Sprintf("%s_%s.jpg", siteDir, details.ScreenshotBase)
		thumb800 := fmt.Sprintf("%s_thumb-800.jpg", details.ScreenshotBase)
		thumb100 := fmt.Sprintf("%s_thumb-100.jpg", details.ScreenshotBase)

		script := fmt.Sprintf(`wget -O %s %s
convert %s -thumbnail 800 -gravity North -crop 800x500+0+0 %s
convert %s -thumbnail 100 %s
rclone move %s %s --fast-list --transfers=32 --no-update-modtime --progress
rclone move %s %s --fast-list --transfers=32 --no-update-modtime --progress
rm %s`,
			downloadFile, sourceURL,
			downloadFile, thumb800,
			thumb800, thumb100,
			thumb100, rcloneDest,
			thumb800, rcloneDest,
			downloadFile,
		)

		shellCmd := exec.Command("bash", "-c", script)
		shellCmd.Dir = system.PathTmp
		shellCmd.Stdout = os.Stdout
		shellCmd.Stderr = os.Stderr
		shellCmd.Run()
	}
}

func init() {
	rootCmd.AddCommand(regenerateThumbnailsCmd)
}
